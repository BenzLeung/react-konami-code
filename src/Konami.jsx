const React = require('react')

class Konami extends React.Component {
  constructor(props) {
    super(props)

    this.state = {
      done: false,
      input: [],
    }
  }

  componentDidMount() {
    document.addEventListener('keyup', this.onKeyUp)
    const delay = Number(this.props.resetDelay)
    if (delay !== 0) {
      this._timer = new this.Timer(() => this.resetInput(), delay)
    }
  }

  onKeyUp = e => {
    const { done, input } = this.state
    const { action, code, disabled, onTimeout, timeout } = this.props
    const resetDelay = Number(this.props.resetDelay)

    if (disabled) {
      return
    }

    if (resetDelay !== 0) {
      this._timer.reset(resetDelay)
    }

    input.push(e.keyCode)
    input.splice(-code.length - 1, input.length - code.length)

    this.setState({ input }, () => {
      if (this.state.input.join('').includes(code.join('')) && !done) {
        if (resetDelay !== 0) {
          this._timer.stop()
        }
        this.setState({ done: true }, () => { 
          action && action()
        })

        if (timeout) {
          setTimeout(() => {
            this.setState({ done: false })
            onTimeout && onTimeout()
          }, Number(timeout))
        }
      }
    })    
  }

  resetInput = () => this.setState({ input: [] })

  Timer(fn, t) {
    let timerObj = setInterval(fn, t)

    this.stop = function() {
      if (timerObj) {
          clearInterval(timerObj)
          timerObj = null
      }
      return this
    }

    this.start = function() {
      if (!timerObj) {
          this.stop()
          timerObj = setInterval(fn, t)
      }
      return this
    }

    this.reset = function(newT) {
      t = newT
      return this.stop().start()
    }
  }

  render = () => {
    const { className, disabled } = this.props
    const { done } = this.state

    return (
      <div 
        className={`konami ${className}`}
        style={{ display: (!done || disabled) ? 'none' : 'block' }}
      >
        {this.props.children}
      </div>
    )
  }
}

Konami.defaultProps = {
  action: null,
  className: "",
  code: [38,38,40,40,37,39,37,39,66,65],
  disabled: false,
  resetDelay: 1000,
  timeout: null,
}

module.exports = Konami