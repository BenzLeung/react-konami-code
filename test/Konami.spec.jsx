import React from 'react';
import Konami from '../src';

describe('Test cases', () => {
  it('should render default component', () => {
    const wrapper = shallow(<Konami>{'Hey, I\'m an Easter Egg!'}</Konami>);
    expect(wrapper).toMatchSnapshot();
  });

  it('should add a className to the wrapping div', () => {
    const wrapper = shallow(<Konami className="customClass">{'Hey, I\'m an Easter Egg!'}</Konami>);
    expect(wrapper).toMatchSnapshot();
  });

  it('should not be done after wrong code input', () => {
    const wrapper = shallow(<Konami>{'Hey, I\'m an Easter Egg!'}</Konami>);

    const instance = wrapper.instance();
    instance.onKeyUp({ keyCode: 38 });
    instance.onKeyUp({ keyCode: 37 });
    instance.onKeyUp({ keyCode: 40 });
    instance.onKeyUp({ keyCode: 40 });
    instance.onKeyUp({ keyCode: 37 });
    instance.onKeyUp({ keyCode: 39 });
    instance.onKeyUp({ keyCode: 37 });
    instance.onKeyUp({ keyCode: 39 });
    instance.onKeyUp({ keyCode: 66 });
    instance.onKeyUp({ keyCode: 65 });

    expect(wrapper.state('done')).toEqual(false);
  });

  it('should be done after code input', () => {
    const cb = jest.fn();
    const wrapper = shallow(<Konami action={cb}>{'Hey, I\'m an Easter Egg!'}</Konami>);

    const instance = wrapper.instance();
    instance.onKeyUp({ keyCode: 38 });
    instance.onKeyUp({ keyCode: 38 });
    instance.onKeyUp({ keyCode: 40 });
    instance.onKeyUp({ keyCode: 40 });
    instance.onKeyUp({ keyCode: 37 });
    instance.onKeyUp({ keyCode: 39 });
    instance.onKeyUp({ keyCode: 37 });
    instance.onKeyUp({ keyCode: 39 });
    instance.onKeyUp({ keyCode: 66 });
    instance.onKeyUp({ keyCode: 65 });

    expect(wrapper.state('done')).toEqual(true);
    expect(cb.mock.calls.length).toEqual(1);
  });

  it('should not be done after code input while disabled', () => {
    const wrapper = shallow(<Konami disabled>{'Hey, I\'m an Easter Egg!'}</Konami>);

    const instance = wrapper.instance();
    instance.onKeyUp({ keyCode: 38 });
    instance.onKeyUp({ keyCode: 38 });
    instance.onKeyUp({ keyCode: 40 });
    instance.onKeyUp({ keyCode: 40 });
    instance.onKeyUp({ keyCode: 37 });
    instance.onKeyUp({ keyCode: 39 });
    instance.onKeyUp({ keyCode: 37 });
    instance.onKeyUp({ keyCode: 39 });
    instance.onKeyUp({ keyCode: 66 });
    instance.onKeyUp({ keyCode: 65 });

    expect(wrapper.state('done')).toEqual(false);
  });

  it('should be done after custom code input', () => {
    const wrapper = shallow(<Konami code={[55, 55]}>{'Hey, I\'m an Easter Egg!'}</Konami>);

    const instance = wrapper.instance();
    instance.onKeyUp({ keyCode: 55 });
    instance.onKeyUp({ keyCode: 55 });

    expect(wrapper.state('done')).toEqual(true);
  });

  it('should not be done after timeout', (done) => {
    const onEnd = jest.fn();
    const wrapper = shallow(
      <Konami code={[55, 55]} onTimeout={onEnd} timeout={1000}>
        {'Hey, I\'m an Easter Egg!'}
      </Konami>,
    );

    const instance = wrapper.instance();
    instance.onKeyUp({ keyCode: 55 });
    instance.onKeyUp({ keyCode: 55 });

    setTimeout(() => {
      expect(wrapper.state('done')).toEqual(false);
      expect(onEnd.mock.calls.length).toEqual(1);
      done();
    }, 1000);
  });

  it('should not be done if you are slower than resetDelay', (done) => {
    const wrapper = shallow(<Konami code={[55, 55]}>{'Hey, I\'m an Easter Egg!'}</Konami>);

    const instance = wrapper.instance();
    instance.onKeyUp({ keyCode: 55 });

    setTimeout(() => {
      instance.onKeyUp({ keyCode: 55 });
      expect(wrapper.state('done')).toEqual(false);
      done();
    }, 1000);
  });

  it('should be done if you are slow but resetDelay is zero', (done) => {
    const wrapper = shallow(<Konami code={[55, 55]} resetDelay={0}>{'Hey, I\'m an Easter Egg!'}</Konami>);

    const instance = wrapper.instance();
    instance.onKeyUp({ keyCode: 55 });

    setTimeout(() => {
      instance.onKeyUp({ keyCode: 55 });
      expect(wrapper.state('done')).toEqual(true);
      done();
    }, 1000);
  });

  it('should reset input if user is slow', (done) => {
    const wrapper = shallow(<Konami code={[55, 55]}>{'Hey, I\'m an Easter Egg!'}</Konami>);

    const instance = wrapper.instance();
    instance.onKeyUp({ keyCode: 55 });

    setTimeout(() => {
      expect(wrapper.state('input')).toEqual([]);
      done();
    }, 1000);
  });

  it('should unmount component', () => {
    const wrapper = shallow(<Konami code={[55, 55]}>{'Hey, I\'m an Easter Egg!'}</Konami>);

    const instance = wrapper.instance();
    const componentWillUnmount = jest.spyOn(instance, 'componentWillUnmount');
    wrapper.unmount();
    expect(componentWillUnmount).toHaveBeenCalled();
  });
});
