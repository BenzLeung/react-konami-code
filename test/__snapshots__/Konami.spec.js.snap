// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test suites should add a className to the wrapping div 1`] = `
<div
  className="konami customClass"
  style={
    Object {
      "display": "none",
    }
  }
>
  Hey, I'm an Easter Egg!
</div>
`;

exports[`test suites should render default component 1`] = `
<div
  className="konami "
  style={
    Object {
      "display": "none",
    }
  }
>
  Hey, I'm an Easter Egg!
</div>
`;
